### PR TITLE
start work on test interpreter

### PR DESCRIPTION
**Pull request by tballmsft**

### CODE REVIEW

File: docs/src/components/ServiceTest.tsx [hunk 1]
```diff
@@ -111,31 +114,22 @@ export default function ServiceTest(props: {
     const { classIdentifier: serviceClass } = serviceSpec
     const [selectedService, setSelectedService] = useState<JDService>(undefined)
     const [activeStep, setActiveStep] = useState(0)
-    const [activeTest, setActiveTest] = useState(-1)
-    const [testStatuses, setTestStatuses]  = useState<TestStatus[]>([])
+    const factory = useCallback(service => new ServiceTestRunner(serviceTest, service),[])
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:117]: [] => [serviceTest]

File: docs/src/components/ServiceTest.tsx [hunk 2]
```diff
@@ -160,21 +154,20 @@ export default function ServiceTest(props: {
                                 />
                             </StepContent>
                         </Step>
-                        {serviceTest?.tests.map((test, index) => (
+                        {testRunner?.tests.map((test, index) => (
                             <Step key={index}>
```
Comment (pelikhan) [inline docs/src/components/ServiceTest.tsx:158]: this should be moved into a separate component;'s generate easier when you refactor those.

Comment (tballmsft) [reply docs/src/components/ServiceTest.tsx:158]: I know - one aspect is that this generates a sequence of Step... if I componentize, I need a parent. Just use a <div>?

Comment (pelikhan) [reply docs/src/components/ServiceTest.tsx:158]: Keep the "<Step>" part in the function, and move the invisde of it in a component

File: src/test/testrunner.ts [hunk 3]
```diff
@@ -67,30 +67,27 @@ export class ServiceTestRunner extends JDServiceClient {
...
+        const spec = serviceSpecificationFromClassIdentifier(this.test.serviceClassIdentifier)
+        // collect up all the registers that are read by the unit tests
+        // TODO: problem here with "position"
+        /*
+        this.test.tests.forEach((ut,index) => {
+            this.unitTests.push(new UnitTestRunner(ut, index));
+            ut.registers.forEach(id => {
+                if (this.deviceState.findIndex(r => r.id === id) < 0) {
+                    const pkt = spec.packets.find(p => { console.log(p.identifierName, id); return p.identifierName === id } )
+                    this.deviceState.push({ id: id, reg: new JDRegister(this.service, pkt.identifier)})
```
Comment (pelikhan) [inline src/test/testrunner.ts:86]: don't do new, do ``service.register(pkt.identifier)``

Comment (tballmsft) [reply src/test/testrunner.ts:86]: thanks

File: src/test/testrunner.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    public reset() {
+        this.unitTests = []
+        this._testIndex = 0
+        const spec = serviceSpecificationFromClassIdentifier(
+            this.test.serviceClassIdentifier
+        )
+        // collect up all the registers that are read by the unit tests
+        this.test.tests.forEach((ut, index) => {
+            this.unitTests.push(new JDTestRunner(ut, () => { 
+                if (index > 0 && index < this.test.tests.length) {
```
Comment (pelikhan) [inline src/test/testrunner.ts:210]: don't pass a callback, use an event instead

File: src/test/testrunner.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            }))
+            // TODO: problem here with "position"
+            /*ut.registers.forEach(id => {
+                if (this.deviceState.findIndex(r => r.id === id) < 0) {
+                    const pkt = spec.packets.find(p => { console.log(p.identifierName, id); return p.identifierName === id } )
+                    this.deviceState.push({ id: id, reg: this.service.register(pkt.identifier)})
+                }
+            })*/
+        })
+        this._testIndex = 0
```
Comment (pelikhan) [inline src/test/testrunner.ts:223]: need to wrap this in a property and trigger a change event

File: src/test/testrunner.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        this.unitTests[0].status = JDTestStatus.Ready
```
Comment (pelikhan) [inline src/test/testrunner.ts:224]: currentTest

File: src/test/testrunner.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    public finishTest(status: JDTestStatus) {
+        this.currentTest.finish(status)
+        if (this._testIndex < this.unitTests.length) {
+            this._testIndex++
+             this.unitTests[this._testIndex].status = JDTestStatus.Ready
```
Comment (pelikhan) [inline src/test/testrunner.ts:231]: this.currentTest.status = ...

Comment (pelikhan) [inline src/test/testrunner.ts:231]: don't allow to modify test state, instead add a command on UnitTestRunner

**Review by pelikhan — CHANGES_REQUESTED** [review]

**Pull request merged**